src/mmrelay/plugin_loader.py

```
@@ -1398,6 +1448,43 @@ def _try_checkout_and_pull_ref(
+    Execute a plugin module while keeping ``sys.modules`` consistent.
+
+    Registering the module before execution ensures ``inspect.getfile()`` works
+    for plugin classes during ``BasePlugin`` initialization (tier inference uses
+    class file paths).
+
+    A module-level lock serialises the ``sys.modules`` manipulation so
+    concurrent plugin loading threads cannot leave the global namespace in an
+    inconsistent state.
+    """
+    if spec.loader is None:
+        raise ImportError(f"No loader available for plugin module '{module_name}'")
```

Ruff (0.15.7)

[warning] Avoid specifying long messages outside the exception class

(TRY003)